### PR TITLE
Refactor proxy handling for Composer2

### PR DESCRIPTION
**Pull request by johnstevenson**
First attempt at fixing the issues highlighted in https://github.com/composer/composer/issues/9314

* ProxyHelper is a utility class that process the environment variables. I've split this out from StreamContextFactory so it can be more easily used by the installer.
* ProxyManager uses the ProxyHelper to provide the context options for streams, and the proxy url for curl. It's a singleton class but I couldn't find any other way of doing it.

In the rare cases of a secure proxy, PHP streams will fail in the case of an https request (in a similar way that a missing openssl extension does). The curl handling also takes a request based approach - however curl cannot handle http requests either so it could lead to a situation where using streams would have worked.

### CODE REVIEW

File: src/Composer/Util/Http/CurlDownloader.php [hunk 1]
```diff
@@ -190,6 +200,26 @@ private function initDownload($resolve, $reject, $origin, $url, $options, $copyT
             }
         }
 
+        // Always set CURLOPT_PROXY to enable/disable proxy handling
+        // Any proxy authorization is included in the proxy url
+        $proxyUrl = $this->proxyManager->getCurloptProxy($url, $secureProxy);
```
Comment (Seldaek) [inline src/Composer/Util/Http/CurlDownloader.php:205]: Could this use `list($proxyUrl, $isSecureProxy) = ...` kind of pattern instead of references? References are really not helping readability, it might make sense when you wrote the code but here I just have to guess at what is going on..

File: src/Composer/Util/Http/CurlDownloader.php [hunk 2]
```diff
@@ -206,8 +236,8 @@ private function initDownload($resolve, $reject, $origin, $url, $options, $copyT
             'reject' => $reject,
         );
 
-        $usingProxy = !empty($options['http']['proxy']) ? ' using proxy ' . $options['http']['proxy'] : '';
-        $ifModified = false !== stripos(implode(',', $options['http']['header']), 'if-modified-since:') ? ' if modified' : '';
+        $usingProxy = $this->proxyManager->getLastProxy(' using proxy (%s)');
+        $ifModified = false !== strpos(strtolower(implode(',', $options['http']['header'])), 'if-modified-since:') ? ' if modified' : '';
```
Comment (Seldaek) [inline src/Composer/Util/Http/CurlDownloader.php:240]: Why the change from stripos? This doesn't seems like an improvement?

File: src/Composer/Util/ProxyManager.php [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+        $info = array();
+
+        if ($httpProxy) {
+            $this->setData($httpProxy, 'http', $info);
+        }
+        if ($httpsProxy) {
+            $this->setData($httpsProxy, 'https', $info);
+        }
+        if ($this->hasProxy) {
+            $this->info = implode(', ', $info);
```
Comment (Seldaek) [inline src/Composer/Util/ProxyManager.php:143]: Generally here too this is super hard to read.. I have no idea what's in $info, I look up and see it's again some by ref argument.. Why isn't setData returning the $info data? Why is it called $info if it represents the safe proxy url? `$safeProxyUrl = $this->setData(...)` would be clearer I'd think.

File: src/Composer/Util/Http/CurlDownloader.php [hunk 4]
```diff
@@ -206,8 +236,8 @@ private function initDownload($resolve, $reject, $origin, $url, $options, $copyT
             'reject' => $reject,
         );
 
-        $usingProxy = !empty($options['http']['proxy']) ? ' using proxy ' . $options['http']['proxy'] : '';
-        $ifModified = false !== stripos(implode(',', $options['http']['header']), 'if-modified-since:') ? ' if modified' : '';
+        $usingProxy = $this->proxyManager->getLastProxy(' using proxy (%s)');
```
Comment (Seldaek) [inline src/Composer/Util/Http/CurlDownloader.php:239]: I was fine with a singleton but now it looks like it's a stateful one too, which makes things really hard to follow and bug prone IMO.

How about converting ProxyManager to something like a ProxyConfig class, which an instance of is created for every request here, it can have a static var for the constant config/state stuff so it doesn't need to recompute it all on every request, but IMO it'd make things much easier, would allow it to have simple getters for values which are sure to be linked to the current URL being processed, no state changes, no by-reference args.

File: src/Composer/Util/ProxyHelper.php [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+        }
+
+        if (isset($parts['user'])) {
+            $parts['user'] = '***';
+        }
+        if (isset($parts['pass'])) {
+            $parts['pass'] = '***';
+        }
+
+        return self::formatParsedUrl($parts, $includeAuth = true);
```
Comment (Seldaek) [inline src/Composer/Util/ProxyHelper.php:80]: I think this whole method could be replaced by `Composer\Util\Url::sanitize($proxyUrl)`

**Review by Seldaek — CHANGES_REQUESTED** [review]

**Comment by Seldaek** [comment]
Also please move those new Proxy* classes into Composer\Util\Http namespace I think it'd make more sense as they are internal.

File: src/Composer/Util/ProxyManager.php (hunk 3 above)
Comment (johnstevenson) [reply src/Composer/Util/ProxyManager.php:143]: Yeah, it's rubbish . I had to remind myself what it is (which is a scheme=safeUrl pair). I'm not sure it's even needed now.

File: src/Composer/Util/Http/CurlDownloader.php (hunk 2 above)
Comment (johnstevenson) [reply src/Composer/Util/Http/CurlDownloader.php:240]: Hmm, I must have screwed up a rebase merge confilct (I hadn't realised this line had change)

File: src/Composer/Util/Http/CurlDownloader.php (hunk 1 above)
Comment (johnstevenson) [reply src/Composer/Util/Http/CurlDownloader.php:205]: Yes, I agree.

File: src/Composer/Util/Http/CurlDownloader.php [hunk 6]
```diff
@@ -190,6 +200,26 @@ private function initDownload($resolve, $reject, $origin, $url, $options, $copyT
...
+        // Always set CURLOPT_PROXY to enable/disable proxy handling
+        // Any proxy authorization is included in the proxy url
+        $proxyUrl = $this->proxyManager->getCurloptProxy($url, $secureProxy);
+        curl_setopt($curlHandle, CURLOPT_PROXY, $proxyUrl);
+
+        // Curl needs certificate locations for secure proxies.
+        // CURLOPT_PROXY_SSL_VERIFY_PEER/HOST are enabled by default
+        if ($proxyUrl && $secureProxy) {
+            if (!$this->supportsSecureProxy) {
+                $message = 'Connecting to a secure proxy using curl is not supported on PHP versions below 7.3.0.';
```
Comment (stof) [inline src/Composer/Util/Http/CurlDownloader.php:212]: why using a variable for the message instead of inlining it ?

File: src/Composer/Util/Http/CurlDownloader.php [hunk 7]
```diff
@@ -176,7 +185,8 @@ private function initDownload($resolve, $reject, $origin, $url, $options, $copyT
         }
 
         $options['http']['header'] = $this->authHelper->addAuthenticationHeader($options['http']['header'], $origin, $url);
-        $options = StreamContextFactory::initOptions($url, $options);
+        // Merge in headers, $forCurl ensures that we don't get any proxy values
+        $options = StreamContextFactory::initOptions($url, $options, $forCurl = true);
```
Comment (stof) [inline src/Composer/Util/Http/CurlDownloader.php:189]: why defining a `$forCurl` local variable here ? That's useless as you don't use it.

File: src/Composer/Util/Http/CurlDownloader.php (hunk 4 above)
Comment (johnstevenson) [reply src/Composer/Util/Http/CurlDownloader.php:239]: Yes that makes more sense: perhaps the ProxyManager could become a ProxyFactory, which stores the state and creates a ProxyConfig (or something) class per request. I originally designed it to be session based, with the HttpDownloader determining if curl is compatible - and using streams if it isn't.

A properly implemented request based approach will make things cleaner - with the proviso that folks using a secure-proxy need PHP 7.3+ and curl (and won't be able to make http requests any more).

File: src/Composer/Util/ProxyHelper.php (hunk 5 above)
Comment (johnstevenson) [reply src/Composer/Util/ProxyHelper.php:80]: Will do.

File: src/Composer/Util/Http/CurlDownloader.php (hunk 6 above)
Comment (johnstevenson) [reply src/Composer/Util/Http/CurlDownloader.php:212]: Agree, there is no need for that (the message was a bit more complicated at one point involving curl versions, until I discovered that even PHP 5.6 supports the version but not the feature)

File: src/Composer/Util/Http/CurlDownloader.php (hunk 7 above)
Comment (johnstevenson) [reply src/Composer/Util/Http/CurlDownloader.php:189]: To make a boolean flag more descriptive and the comment easier to write/understand, but I can remove it if you want?

Comment (stof) [reply src/Composer/Util/Http/CurlDownloader.php:189]: But then, phpstan (and IDEs) would complain about an unused variable (phpstan does not complain right now because composer runs it at a very low level and this rule is at a higher level).
Regarding the behavior of `$forCurl == true`, the proper place to document its effect would be in the phpdoc of `StreamContextFactory::initOptions` IMO, not on the usage

File: src/Composer/Util/ProxyHelper.php (hunk 5 above)
Comment (johnstevenson) [reply src/Composer/Util/ProxyHelper.php:80]: Oops, that last reply was intended to be to "Also please move those new Proxy* classes into Composer\Util\Http namespace I think it'd make more sense as they are internal."

I had hoped that it would do the job, but it doesn't work without a scheme. Incidentally, the ProxyHelper version masks the user but Url::sanitize does not - so is it safe (good practice) to show the user part ?

Comment (Seldaek) [reply src/Composer/Util/ProxyHelper.php:80]: Well generally speaking showing the user can help people spot misconfiguration I'd think. Feel free to tweak the sanitize method so it handles URLs without scheme tho. I guess turning `://` into `(://|^)` in the regex and then returning $m[1] instead of :// should do.

File: src/Composer/Util/Http/CurlDownloader.php (hunk 7 above)
Comment (johnstevenson) [reply src/Composer/Util/Http/CurlDownloader.php:189]: Thanks for explaining, I hadn't realised the phpstan bit.

Comment (johnstevenson) [reply src/Composer/Util/Http/CurlDownloader.php]: Resolved in Code review:1 fixes

File: src/Composer/Util/Http/CurlDownloader.php (hunk 2 above)
Comment (johnstevenson) [reply src/Composer/Util/Http/CurlDownloader.php]: Resolved in Code review:1 fixes

File: src/Composer/Util/ProxyHelper.php (hunk 5 above)
Comment (johnstevenson) [reply src/Composer/Util/ProxyHelper.php]: Resolved in Code review:1 fixes

File: src/Composer/Util/Http/CurlDownloader.php (hunk 6 above)
Comment (johnstevenson) [reply src/Composer/Util/Http/CurlDownloader.php]: Resolved in Code review:1 fixes

File: src/Composer/Util/Http/CurlDownloader.php (hunk 1 above)
Comment (johnstevenson) [reply src/Composer/Util/Http/CurlDownloader.php]: Resolved in Refactor ProxyManager

File: src/Composer/Util/Http/CurlDownloader.php (hunk 4 above)
Comment (johnstevenson) [reply src/Composer/Util/Http/CurlDownloader.php]: Resolved in Refactor ProxyManager

File: src/Composer/Util/ProxyManager.php (hunk 3 above)
Comment (johnstevenson) [reply src/Composer/Util/ProxyManager.php]: Resolved in Refactor ProxyManager

**Comment by johnstevenson** [comment]
@Seldaek Okay, we're almost there. I need to add tests to the `RequestProxy` class (terrible name) but the core data is already covered so these can wait.

I'm going to spend a short time running it through my secure-proxy implementation, then I'll rebase it all. Shout now if you want anything changing or spot something that is glaringly awful or just plain wrong.

**Comment by johnstevenson** [comment]
Squashed and rebased. It appears to work well running it through a proxy.

File: src/Composer/Util/Http/CurlDownloader.php [hunk 8]
```diff
@@ -176,7 +184,8 @@ private function initDownload($resolve, $reject, $origin, $url, $options, $copyT
         }
 
         $options['http']['header'] = $this->authHelper->addAuthenticationHeader($options['http']['header'], $origin, $url);
-        $options = StreamContextFactory::initOptions($url, $options);
+        // Merge in headers - we don't get any proxy values
```
Comment (Seldaek) [inline src/Composer/Util/Http/CurlDownloader.php:187]: Not sure I understand what this means.. We don't get any proxy values? And what merging of headers?

File: src/Composer/Util/StreamContextFactory.php [hunk 9]
```diff
@@ -70,76 +72,26 @@ public static function initOptions($url, array $options)
...
-                    break;
+        // Add stream proxy options if there is a proxy
+        if (!$forCurl) {
+            $proxy = ProxyManager::getInstance()->getProxyForRequest($url);
+            if ($proxy->isSecure()) {
+                if (!extension_loaded('openssl')) {
+                    throw new TransportException('You must enable the openssl extension to use a proxy over https.');
+                }
+                if (0 === strpos($url, 'https://')) {
+                    throw new TransportException('PHP does not support https requests to a secure proxy.');
```
Comment (Seldaek) [inline src/Composer/Util/StreamContextFactory.php:83]: Maybe suggest enabling the curl extension, and specify that PHP **streams** do not support this? Generally speaking the more verbose and helpful the errors are the better, if they allow people to fully understand or at least fix the problem without reaching out to us.

**Review by Seldaek — APPROVED** [review]
Looks fine overall so merging, still left a couple comments and fixed a couple things myself in 62eff8e979a5adeffc2b5490f581d5ea5bbf008e
